bcm27xx: add USB boot support to Raspberry Pi's

michaelr0 opened the pull request:
This PR aims to add USB boot support to Raspberry Pi.

I have had success with USB booting a Raspberry Pi 4 Model B into OpenWrt, however this may need to serve as a discussion as I'm sure that there might be more to OpenWrt supporting USB booting on a Pi, than simply changing /dev/mmcblk0p2 to the PARTUUID=5452574f-02 in /boot/cmdline.txt 

I can see that there are is a reference to `/dev/mmcblk0p1` in [79_move_config](https://github.com/openwrt/openwrt/blob/master/target/linux/bcm27xx/base-files/lib/preinit/79_move_config#L5) which might be able to be changed to `5452574f-01`? but feel free to chime in as I'm not that familar with OpenWrt.

Thanks.

Signed-off-by: Michael Rook <michael@rook.net.au>

Entropy512 commented:
Yeah, you definitely need that change in 79_move_config

Good job on figuring out a UUID mechanism that plays nice with the Pi bootloader

What I'm not sure about is whether or not that UUID is reliable/consistent, although it DOES appear that it is correct for USB in my current builds.  I've been doing USB boot by changing from mmcblk0pN to sdaN in the two locations needed.

Will https://github.com/openwrt-mirror/openwrt/blob/master/tools/firmware-utils/src/ptgen.c generate a table with a different UUID if someone changes the bootfs size or rootfs size?  (It looks like, by default, it will - https://github.com/openwrt-mirror/openwrt/blob/master/tools/firmware-utils/src/ptgen.c#L198 ?)

walterav1984 commented:
Are both USB 2 and 3 ports supported (ehci/xhci) booting both USB2&3 viceversa master storage devices? 

No need for other kernel arguments like `rootdelay=15` `bootwait`, than its kinda surprise that in 2022 devices still get booted by block device number than UUID.

Entropy512 commented:
Both USB2 and USB3 can be booted from, I confirmed USB2 on a pi that had a glitchy USB3 port.

I have not found a way to rework 79_move_config in a way that it actually works with PARTUUID yet...  unfortunately preinit debugging is voodoo black magic.

walterav1984 commented:
Did the image still boot from sdcard if using UUID instead of `/dev/mmcblk0p1` than it might be a very simple fix?

Entropy512 commented:
The problem is not that the image doesn't boot - it's that 79_move_config won't mount /boot, and that causes the image to boot into a default configuration (e.g. configuration is not preserved)

It looks like currently 79_move_config is hardcoded to check if the block device exists before trying to mount /boot - I'm testing an approach that steals a bunch from https://github.com/openwrt/openwrt/blob/master/target/linux/imx/cortexa9/base-files/lib/imx.sh#L17 now

Entropy512 commented:
https://gist.githubusercontent.com/Entropy512/50e0d2b6d4078855e28a736c95486767/raw/cd6a4f73a32c9c171e0ae910efc6298983438bd2/0001-bcm27xx-Rework-79_move_config-to-use-PARTUUID.patch

michaelr0 commented:
> Will https://github.com/openwrt-mirror/openwrt/blob/master/tools/firmware-utils/src/ptgen.c generate a table with a different UUID if someone changes the bootfs size or rootfs size? (It looks like, by default, it will - https://github.com/openwrt-mirror/openwrt/blob/master/tools/firmware-utils/src/ptgen.c#L198 ?)

I guess it depends on if it changes the UUID and or the PARTUUID, as it is my understanding that although these might have the same values, they are two separate identifiers? And that PARTUUID shouldn’t change, while UUID probably will?

This should mount correctly,
```
mount -t vfat -o rw,noatime PARTUUID=5452574f-01 /boot
```

So the move config might be as simple as changing 

````
BOOTPART=/dev/mmcblk0p1
```

to

```
BOOTPART=PARTUUID=5452574f-01
```

michaelr0 commented:
> Did the image still boot from sdcard if using UUID instead of `/dev/mmcblk0p1` than it might be a very simple fix?

Yeah the SD card contains the same PARTUUID’s and I’ve tried both USB and SD card on my Pi 4 and both USB 3 and 2 function for me.

As far as as the boot order on a Pi 4 and a Pi Zero 2 W, after you enable USB booting, it’ll try booting from USB and if it fails to find a bootable target it’ll try to boot from SD be so fallback options are possible I guess?

Entropy512 commented:
@michaelr0 I tried exactly that (just changing BOOTPART), and it fails because the script is checking to see if BOOTPART is a block device.

See the gist that I posted - it's a patch (you should be able to add it to your branch using 'git am') that does the following:
Adds blkid to DEFAULT_PACKAGES (needed for the reworked script)
Steals a bunch of partition ID stuff from the i.MX6 support in OpenWRT.

And yeah, by default newer Pi 4 bootloaders try SD and then fallback to USB (or is the default the other way around?), once booted, PARTUUID should work as long as you don't have two different OpenWRT builds on two different storage devices.  That might break - in the future there's some additional stuff in either the iMX or x86 support trees that dynamically generates a PARTUUID signature from the git revision and build time, instead of forcing it to be the hexadecimal representation of OWRT in ASCII, and then puts that signature into the config files.

@walterav1984 FYI, the Raspi bootloader configuration documentation is at https://www.raspberrypi.com/documentation/computers/raspberry-pi.html#raspberry-pi-4-bootloader-configuration - openWRT does have the rpi-eeprom-config utility as an available package (I forget the exact package name), so you can update the bootloader and reconfigure it from within OpenWRT.  That's basically how I moved a pile of old Pi 4s to USB boot - I booted OpenWRT from SD, and used rpi-eeprom-update and rpi-eeprom-config to update bootloaders and configure them to boot from USB and fall back to SD if that failed.

michaelr0 commented:
Hi @Entropy512,

I'm currently in the process of building a sysupgrade image to test out upgrading, I'm putting that [here](https://github.com/michaelr0/rpi4-openwrt)

Perhaps it is worth including the `findfs` package, which is 3.2 KB, which should allow the mount command to get the device path from the PARTUUID.

```bash
root@OpenWrt:~# ls /boot
root@OpenWrt:~# echo $(findfs PARTUUID=5452574f-01)
/dev/sda1
root@OpenWrt:~# mount -t vfat -o rw,noatime $(findfs PARTUUID=5452574f-01) /boot
root@OpenWrt:~# ls /boot
COPYING.linux        bcm2711-rpi-cm4.dtb  fixup4.dat           overlays
LICENCE.broadcom     cmdline.txt          fixup4cd.dat         start4.elf
bcm2711-rpi-4-b.dtb  config.txt           fixup4x.dat          start4cd.elf
bcm2711-rpi-400.dtb  distroconfig.txt     kernel8.img          start4x.elf
root@OpenWrt:~#
```

Which would then make `79_move_config` look something like this.
```
# Copyright (C) 2015 OpenWrt.org

. /lib/upgrade/common.sh

BOOTPART=$(findfs PARTUUID=5452574f-01)

move_config() {
	if [ -b $BOOTPART ]; then
		insmod nls_cp437
		insmod nls_iso8859-1
		insmod fat
		insmod vfat
		mkdir -p /boot
		mount -t vfat -o rw,noatime $BOOTPART /boot
		[ -f "/boot/$BACKUP_FILE" ] && mv -f "/boot/$BACKUP_FILE" /
	fi
}

boot_hook_add preinit_mount_root move_config
```

I haven't yet tested your patch, but looking at the code, it seems plausible :)
```
# Copyright (C) 2015 OpenWrt.org

. /lib/upgrade/common.sh

rootpartuuid() {
        local cmdline=$(cat /proc/cmdline)
        local bootpart=${cmdline##*root=}
        bootpart=${bootpart%% *}
        local uuid=${bootpart#PARTUUID=}
        echo ${uuid%-02}
}

bootpart_from_uuid() {
        blkid | grep $(rootpartuuid)-01 | cut -d : -f1
}

move_config() {
	if [ -b $(bootpart_from_uuid) ]; then
		insmod nls_cp437
		insmod nls_iso8859-1
		insmod fat
		insmod vfat
		mkdir -p /boot
		mount -t vfat -o rw,noatime $(bootpart_from_uuid) /boot
		[ -f "/boot/$BACKUP_FILE" ] && mv -f "/boot/$BACKUP_FILE" /
	fi
}

boot_hook_add preinit_mount_root move_config
```

michaelr0 commented:
> I'm currently in the process of building a sysupgrade image to test out upgrading, I'm putting that [here](https://github.com/michaelr0/rpi4-openwrt)
> 
> Perhaps it is worth including the `findfs` package, which is 3.2 KB, which should allow the mount command to get the device path from the PARTUUID.
> 
> ```shell
> root@OpenWrt:~# ls /boot
> root@OpenWrt:~# echo $(findfs PARTUUID=5452574f-01)
> /dev/sda1
> root@OpenWrt:~# mount -t vfat -o rw,noatime $(findfs PARTUUID=5452574f-01) /boot
> root@OpenWrt:~# ls /boot
> COPYING.linux        bcm2711-rpi-cm4.dtb  fixup4.dat           overlays
> LICENCE.broadcom     cmdline.txt          fixup4cd.dat         start4.elf
> bcm2711-rpi-4-b.dtb  config.txt           fixup4x.dat          start4cd.elf
> bcm2711-rpi-400.dtb  distroconfig.txt     kernel8.img          start4x.elf
> root@OpenWrt:~#
> ```
> 
> Which would then make `79_move_config` look something like this.
> 
> ```
> # Copyright (C) 2015 OpenWrt.org
> 
> . /lib/upgrade/common.sh
> 
> BOOTPART=$(findfs PARTUUID=5452574f-01)
> 
> move_config() {
> 	if [ -b $BOOTPART ]; then
> 		insmod nls_cp437
> 		insmod nls_iso8859-1
> 		insmod fat
> 		insmod vfat
> 		mkdir -p /boot
> 		mount -t vfat -o rw,noatime $BOOTPART /boot
> 		[ -f "/boot/$BACKUP_FILE" ] && mv -f "/boot/$BACKUP_FILE" /
> 	fi
> }
> 
> boot_hook_add preinit_mount_root move_config
> ```

So that seems like it works, although sysupgrade complains about metadata and a signature, so it requires the force option.

Entropy512 commented:
I'll test your version of the patch on Monday sometime.

michaelr0 commented:
> I'll test your version of the patch on Monday sometime.

Cheers

michaelr0 commented:
> I didn't get a chance to test until today.
> 
> Your first patch works, but your second patch causes the fallback to ALWAYS be used even if findfs is present. Either that or BOOTPART is set to something completely invalid, which has the same results - any configuration is obliterated, resetting the device to defaults. Reverting the second patch out leads to things working.
> 
> (Easiest way to see this is to set an IP address other than default - if you see the IP address revert to default, 79_move_config failed)
> 
> I don't currently have any devices booting from SD, which would be able to determine if it's a complete failure or just the fallback always being used.

Thanks @Entropy512, 

I'm going to try wrapping the fallback into an if ... else instead and seeing if that works for USB and Micro SD.

```bash
# If findfs is installed, try to find the BOOTPART via the BOOTPARTUUID
# Allows booting to devices which are not /dev/mmcblk0, such as USB, eMMC.
# If not, then fallback to /dev/mmcblk0p1
BOOTPARTUUID="5452574f-01"
if [ -x $(command -v findfs)]; then
	BOOTPART=$(findfs PARTUUID=$BOOTPARTUUID)
else
        BOOTPART=/dev/mmcblk0p1
fi
```

I'll report back to let you know how that goes.

michaelr0 commented:
> I'm going to try wrapping the fallback into an if ... else instead and seeing if that works for USB and Micro SD.
> 
> ```shell
> # If findfs is installed, try to find the BOOTPART via the BOOTPARTUUID
> # Allows booting to devices which are not /dev/mmcblk0, such as USB, eMMC.
> # If not, then fallback to /dev/mmcblk0p1
> BOOTPARTUUID="5452574f-01"
> if [  -x $(command -v findfs) ]; then
> 	BOOTPART=$(findfs PARTUUID=$BOOTPARTUUID)
> else
>         BOOTPART=/dev/mmcblk0p1
> fi
> ```
> 
> I'll report back to let you know how that goes.

Well, that wasn't very successful.

This appears to work for me though, on both usb and micro sd.
Added to commit: https://github.com/openwrt/openwrt/pull/9843/commits/f4493a33b1f2fbc01a2097ba3a982d1c576dd52d
```bash
# Copyright (C) 2015 OpenWrt.org

. /lib/upgrade/common.sh

# Default SD card partition used for booting
BOOTPART=/dev/mmcblk0p1

# If findfs is installed, try to find the BOOTPART via the BOOTPARTUUID
# This allows booting to devices which are not /dev/mmcblk0, such as USB, eMMC.
# 5452574f-01 is the default PARTUUID for the boot partition.
BOOTPARTUUID="5452574f-01"
if [ -x $(command -v findfs) ]; then
	# findfs appears to be installed, lets try to find the boot partition by PARTUUID
    FINDFSPART=$(findfs PARTUUID=$BOOTPARTUUID)
    if [ $? -eq 0 ]; then
		# findfs was able to find the PARTUUID.
	    BOOTPART=$FINDFSPART
    fi
fi

move_config() {
	if [ -b $BOOTPART ]; then
		insmod nls_cp437
		insmod nls_iso8859-1
		insmod fat
		insmod vfat
		mkdir -p /boot
		mount -t vfat -o rw,noatime $BOOTPART /boot
		[ -f "/boot/$BACKUP_FILE" ] && mv -f "/boot/$BACKUP_FILE" /
	fi
}

boot_hook_add preinit_mount_root move_config
```

michaelr0 commented:
Hi @mans0n,

Anychance you could rerun the `Test Formalities` action? It should be compliant now.

michaelr0 commented:
With some additional digging, I have found that the `platform_copy_config` function in https://github.com/openwrt/openwrt/blob/master/target/linux/bcm27xx/base-files/lib/upgrade/platform.sh#L88 looks to be responsible for the configuration backup that `move_config` tries to restore, and it looks like `platform_copy_config` uses another function called `export_partdevice` to set the `partdev` variable to be the returned path for a particular offset (1 for boot, 2 for root)

Which is interesting, because after finding that and searching for it, I can see that a few other `79_move_config` targets are doing it that way (each with their own little variations)

- [mediatek/mt7623](https://github.com/openwrt/openwrt/blob/master/target/linux/mediatek/mt7623/base-files/lib/preinit/79_move_config)
- [mvebu](https://github.com/openwrt/openwrt/blob/master/target/linux/mvebu/base-files/lib/preinit/79_move_config)
- [omap](https://github.com/openwrt/openwrt/blob/master/target/linux/omap/base-files/lib/preinit/79_move_config)
- [qorig](https://github.com/openwrt/openwrt/blob/master/target/linux/qoriq/base-files/lib/preinit/79_move_config)
- [rockchip/armv8](https://github.com/openwrt/openwrt/blob/master/target/linux/rockchip/armv8/base-files/lib/preinit/79_move_config)
- [sunxi](https://github.com/openwrt/openwrt/blob/master/target/linux/sunxi/base-files/lib/preinit/79_move_config)
- [tegra](https://github.com/openwrt/openwrt/blob/master/target/linux/tegra/base-files/lib/preinit/79_move_config)
- [x86](https://github.com/openwrt/openwrt/blob/master/target/linux/x86/base-files/lib/preinit/79_move_config)

So with that in mind, it is plausible that `79_move_config` could look like the below, and wouldn't need to have the `findfs` package installed either.

```bash
# Copyright (C) 2015 OpenWrt.org

. /lib/upgrade/common.sh

move_config() {
	local partdev

	if export_bootdevice && export_partdevice partdev 1; then
		insmod nls_cp437
		insmod nls_iso8859-1
		insmod fat
		insmod vfat
		mkdir -p /boot
		mount -t vfat -o rw,noatime "/dev/$partdev" /boot
		[ -f "/boot/$BACKUP_FILE" ] && mv -f "/boot/$BACKUP_FILE" /
	fi
}

boot_hook_add preinit_mount_root move_config
```

I have not yet tested this, but will give it a try shortly and report back.

michaelr0 commented:
Update: that works! I'll push those changes now.

Entropy512 commented:
I can confirm that both a Pi 3B+ booting from microSD and a Pi 4B booting from USB are working with your patchset.

You probably want to squash your last three commits into one at the very least.  Possibly even just squash everything into one commit to keep things clean and tidy.

michaelr0 commented:
Thanks @Entropy512,

I'll squash them shortly :)

Commit:
bcm27xx: add USB Boot support to Raspberry Pi's

Add USB boot support to Raspberry Pi targets,
this has been done by changing /boot/cmdline.txt
and /lib/preinit/79_move_config as detailed below.

Change hardcoded rootfs in /boot/cmdline.txt,
now uses PARTUUID=5452574f-02 instead of /dev/mmcblk0p2

Change /lib/preinit/79_move_config to detect the boot partition,
instead of using the previously hardcoded /dev/mmcblk0p1

Signed-off-by: Michael Rook <michael@rook.net.au>

michaelr0 commented:
Hi @mans0n,

Sorry to bug you again, would it be possible to rerun the Test Formalities action?
Also please forgive me, since this is my first submission to Openwrt, do I need to request a review, or will that just happen when the reviewers have availability?

Thanks.

Entropy512 commented:
BTW, I can confirm that your squashed patches have no apparent regressions on the same configurations tested in my last comment.

michaelr0 commented:
> You can just wait for reviewers who might be interested. For me, I feel I'm not the right person to review this, sorry about that.

Thanks @mans0n for your help thus far anyway, with running those actions and setting the labels on the PR.

> BTW, I can confirm that your squashed patches have no apparent regressions on the same configurations tested in my last comment.

Thanks @Entropy512

Entropy512 commented:
@mans0n - you flagged this as needs changes earlier - and IIRC at that time, changes *were* needed.  The "needs changes" flag is still present after significant improvements were made (and I am assuming this is inhibiting attention by reviewers), what is missing now?

mans0n commented:
I don't see any other formal issues, removed it.

michaelr0 commented:
Thanks @mans0n, @Entropy512

michaelr0 commented:
Is there any way to help this move along? is anything outstanding or do I need to do something for this to progress further down the line?

michaelr0 commented:
This looks like its gone into limbo?
> Is there any way to help this move along? is anything outstanding or do I need to do something for this to progress further down the line?

wulfy23 commented:
Thanks for submitting this. Have not tested it but i'm sure it is fundamentally sound.

Touching on a really good point raised by @Entropy512

It may be prudent to add a note in the commit recommending users manually update their cmdline.txt to PARTUUID as currently someone blindly restoring a backup from static a prev mmc install onto a newly flashed usb disk will essentially be rendering their system non-bootable due these boottime confs are bundled into backups.

Or a link to bootorder documentation &&|| cmdline.txt editing on a PC in the event of boot failure or similar. These devices while being 'enthusiast' aka aimed at peeps who are capable/expected to tinker and school themselves on lower level device intricacies often each an audience without those inclinations or skillsets.

Caveat / Nitpick 1
I can forsee someone dd-ing an mmc to usb say from a nfs booted OS or alike. One would assume that everything will roll according to bootorder if two identical disks are installed. But I don't see info if this level of testing has been undertaken. (i.e. does it boot? does it mount the rootfs on the correct disk, does an upgrade apply to the right places, do all proceeding checks also checkout if boot order is switched? Edge case maybe... and likely not a blocker in anyway for this PR... but food for thought none the less...

Caveat 2
Dude/Dudette uses disk tools to expand his disks and it blasts the PARTUUID. Again not a blocker but it will happen. Perhaps something to document on the wiki? (disk expansion guides etc.)

Anyways, once again thanks for the PR and your patience.

michaelr0 commented:
Thanks @wulfy23 

Those are decent concerns, I'm not exactly sure how to approch them or warn about them.
Would you be able to squeeze in some testing of the changes, to see if they also work for you?

wulfy23 commented:
They will work. (unless I resize the disk || keep settings then swap the disk as discussed) or restore an old backup once booted from USB, as discussed. Probably not enough to require config-version bump but definitely enough to warrant mention.

michaelr0 commented:
Just popped by to see how this was going, seen that it needed to be rebased since it was outdated, so I have rebased this now.

Entropy512 commented:
While I still use OpenWRT, I no longer use it with a Pi due to a change of employment, so unfortunately not I...

michaelr0 commented:
@hauke @graysky2 Sorry for tagging you two, but I figure you both seem to be active in development of OpenWrt on Pi's and that you might be the best people for this one?

Thanks and fingers crossed

Noltari commented:
Superseded by https://github.com/openwrt/openwrt/pull/14520
